changelog.lua

```diff
@@ -0,0 +1,374 @@
+    deprecation = "deprecations",
+    performance = "performances",
+  }
+
+  local files = collect_files(sub_folder)
+  for _, filename in ipairs(files) do
+    local content = assert(pl_file.read(filename))
+    local entry = assert(lyaml.load(content))
+
+    if is_empty_table(entry.prs) or is_empty_table(entry.jiras) then
+      local ctx, err = fetch_commit_context(filename)
+      if not ctx then
+        error("faield to fetch commit context: " .. err)
+      end
+      --print(cjson.encode(ctx))
+      if is_empty_table(entry.prs) then
+        entry.prs = { ctx.pull.number }
+      end
+      if is_empty_table(entry.jiras) then
+        local jiras = {}
+
+        local sources = { string.upper(ctx.commit.message), string.upper(ctx.pull.body) }
+        for _, text in ipairs(sources) do
+          match_pattern(text, "FTI%-%d+", jiras)
+          match_pattern(text, "KAG%-%d+", jiras)
+        end
+
+        entry.jiras = {}
+        for _, jira in ipairs(jiras) do
+          if not jiras[jira] then
+            table.insert(entry.jiras, jira)
+            jiras[jira] = true
+          end
+        end
+      end
+    end
+
+    entry.prs = parse_github_ref(system, "pr", entry.prs) or {}
+    entry.issues = parse_github_ref(system, "issue", entry.issues) or {}
+    entry.jiras = parse_jiras(entry.jiras) or {}
+
+
+    if entry.scope == nil or entry.scope == lyaml.null then
+      entry.scope = "Default"
+    end
+
+    local key = map[entry.type]
+    if not data[key][entry.scope] then
+      data[key][entry.scope] = {}
+    end
+    table.insert(data[key][entry.scope], entry)
+  end
+
+  for _, scopes in pairs(data) do
+    local scope_names = table_keys(scopes)
+    table.sort(scope_names, function(a, b) return SCOPE_PRIORITY[a] < SCOPE_PRIORITY[b] end)
+    scopes.sorted_scopes = scope_names
+  end
+
+  return data
+end
+
+local function collect_folder(path, folder)
+  if not pl_path.exists(pl_path.join(path, folder)) then
+    error("folder: " .. pl_path.join(path, folder) .. " does not exist")
+  end
+
+  local data = {}
+
+  data.kong = collect_sub_folder("kong", pl_path.join(path, folder, "kong"))
+
+  return data
+end
+
+
+local function generate_content(data)
+  local template_path = absolute_path("changelog-md-template.lua")
+  local content = assert(pl_file.read(template_path))
+  local changelog_template = assert(loadstring(content))()
+  return compile_template(data, changelog_template)
+end
+
+
+-- command: release
+-- release a release note
+local function release(version, path, folder)
+  local data = collect_folder(path, folder)
+  data.version = version
+  local content = assert(generate_content(data))
+  local target_path = pl_path.join(path, version)
+
+  if pl_path.exists(target_path) and version ~= folder then
+    error("directory exists, please manually remove. " .. version)
+  end
+
+  if folder == UNRELEASED then
+    os.execute("mv " .. pl_path.join(path, UNRELEASED) .. " " .. target_path)
```

Comment:
These does not really need to be so automatic, I would not feel too comfortable having a script `mv` file on the disk for me, it is very easy to make a mistake in the command and cause damage. 

Just print out the command and let user execute it themself.

I even argue this script should not even write the `.md` file directly, use the KISS principal and let user do:

`./changelog.lua 3.4.0. /path/to/kong/unreleased > CHANGELOG-3.4.0.md`

is much better and safer.

If we really need to automate, Makefile and a shell script is much better idea.